SmartDeviceLinkTests/ProtocolSpecs/SDLProtocolReceivedMessageProcessorSpec.m

```diff
@@ -0,0 +1,736 @@
+    MESSAGE_2_STATE = 0x09,
+    MESSAGE_3_STATE = 0x0A,
+    MESSAGE_4_STATE = 0x0B,
+    DATA_PUMP_STATE = 0x0C,
+    ERROR_STATE = -1,
+};
+
+@interface SDLProtocolReceivedMessageProcessor ()
+// State management
+@property (assign, nonatomic) ProcessorState state;
+
+// Message assembly state
+@property (strong, nonatomic) SDLProtocolHeader *header;
+@property (strong, nonatomic) NSMutableData *headerBuffer;
+@property (strong, nonatomic) NSMutableData *payloadBuffer;
+
+// Error checking
+@property (assign, nonatomic) UInt8 version;
+@property (assign, nonatomic) BOOL encrypted;
+@property (assign, nonatomic) SDLFrameType frameType;
+@property (assign, nonatomic) UInt32 dataLength;
+@property (assign, nonatomic) UInt32 dataBytesRemaining;
+@property (assign, nonatomic) SDLServiceType serviceType;
+@end
+
+QuickSpecBegin(SDLProtocolReceivedMessageProcessorSpec)
+
+describe(@"The received message processor", ^{
+    __block SDLProtocolReceivedMessageProcessor *testProcessor = nil;
+    __block NSMutableData *testBuffer;
+    __block NSMutableData *testHeaderBuffer;
+    __block SDLProtocolHeader *messageReadyHeader = nil;
+    __block SDLProtocolHeader *expectedMessageReadyHeader = nil;
+    __block NSData *messageReadyPayload = nil;
+    __block NSData *expectedPayloadBuffer = nil;
+    
+    beforeEach(^{
+        testProcessor = [[SDLProtocolReceivedMessageProcessor alloc] init];
+        testBuffer = [NSMutableData data];
+        testHeaderBuffer = [NSMutableData data];
+        messageReadyHeader = nil;
+        expectedMessageReadyHeader = nil;
+        messageReadyPayload = nil;
+    });
+
+    it(@"test processor should be initialized correctly", ^{
+        expect(@(testProcessor.state)).to(equal(START_STATE));
+        expect(@(testProcessor.version)).to(equal(0));
+        expect(@(testProcessor.encrypted)).to(equal(NO));
+        expect(@(testProcessor.frameType)).to(equal(SDLFrameTypeControl));
+        expect(@(testProcessor.dataLength)).to(equal(0));
+        expect(@(testProcessor.dataBytesRemaining)).to(equal(0));
+        expect(@(testProcessor.serviceType)).to(equal(SDLServiceTypeControl));
+    });
+    
+    describe(@"transitions to SERVICE_TYPE_STATE when in START_STATE", ^{
+
+        it(@"should receive a byte with a good version 1", ^{
+            Byte testByte = 0x11;
+            [testBuffer appendBytes:&testByte length:1];
+
+            expect(messageReadyPayload).toEventually(beNil());
+            [testProcessor processReceiveBuffer:testBuffer withMessageReadyBlock:^(SDLProtocolHeader *header, NSData *payload) {
+                messageReadyHeader = header;
+                messageReadyPayload = payload;
+                expect(messageReadyPayload).toEventually(beNil());
+            }];
+            expect(@(testProcessor.state)).to(equal(SERVICE_TYPE_STATE));
+            expect(messageReadyHeader).toEventually(beNil());
+            expect(messageReadyPayload).toEventually(beNil());
```

Comment:
```suggestion
                expect(messageReadyHeader).to(beNil());
                expect(messageReadyPayload).to(beNil());
            }];
            expect(@(testProcessor.state)).to(equal(SERVICE_TYPE_STATE));
```

Because of how tests work, you don't usually want completion handlers to have `expect`s. Race conditions can cause them to not be called and then fail. **However,** because this completion handler is called synchronously, I *think* we can put the tests directly in the handler. Make sure they're called properly and update the other tests, please.